core/src/HIP/Kokkos_HIP_Space.cpp

```diff
@@ -810,3 +551,26 @@ void HIPSpaceInitializer::print_configuration(std::ostream& msg,
 
 }  // namespace Impl
 }  // namespace Kokkos
+
+//==============================================================================
+// <editor-fold desc="Explicit instantiations of CRTP Base classes"> {{{1
+
+#include <impl/Kokkos_SharedAlloc_timpl.hpp>
```

Comment:
I agree. This isn't in the middle of the file, though; it's at the end. There are a lot of reasons to do this specifically for template instantiation files (`*_timpl.hpp` is the pattern I've been using for this, but we can use something else if you'd prefer). One good one is to make sure that the rest of your code doesn't depend in some way on the explicit instantiations of these templates; another is to make sure any partial specializations specific to the current implementation file get seen by the template instantiation code. A lot of people put these instantiations in their own file for these reasons, but I didn't want to mess with the build system on that front (I'm sure it would have been fine, but the "template instantiations and their include files go at the end of the implementation file" rule is an okay one also in my book, so I just went with that).

Reply:
Fair enough. In the end, this is not a header file.